remote-wallet/src/remote_wallet.rs

```diff
@@ -250,65 +254,234 @@ pub struct RemoteWalletInfo {
+pub enum LocaterError {
+    #[error(transparent)]
+    VendorError(#[from] VendorError),
+    #[error(transparent)]
+    PubkeyError(#[from] ParsePubkeyError),
+    #[error(transparent)]
+    DerivationPathError(#[from] DerivationPathError),
+    #[error(transparent)]
+    UriReferenceError(#[from] URIReferenceError),
+    #[error("unimplemented scheme")]
+    UnimplementedScheme,
+    #[error("infallible")]
+    Infallible,
+}
+
+impl From<Infallible> for LocaterError {
+    fn from(_: Infallible) -> Self {
+        Self::Infallible
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Locater {
+    vendor: Vendor,
+    pubkey: Option<Pubkey>,
+    derivation_path: Option<DerivationPath>,
+}
+
+impl std::fmt::Display for Locater {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        let maybe_path = self.pubkey.map(|p| p.to_string());
+        let path = maybe_path.as_deref().unwrap_or("/");
+        let maybe_query = self.derivation_path.as_ref().map(|d| d.get_query());
+        let maybe_query2 = maybe_query.as_ref().map(|q| &q[1..]);
+
+        let mut builder = URIReferenceBuilder::new();
+        builder
+            .try_scheme(Some("usb"))
+            .unwrap()
+            .try_authority(Some(self.vendor.as_ref()))
+            .unwrap()
+            .try_path(path)
+            .unwrap()
+            .try_query(maybe_query2)
+            .unwrap();
+
+        let uri = builder.build().unwrap();
+        write!(f, "{}", uri)
+    }
+}
+
+impl Locater {
+    pub fn new_from_path<P: AsRef<str>>(path: P) -> Result<Self, LocaterError> {
+        let path = path.as_ref();
+        let uri = URIReference::try_from(path)?;
+        Self::new_from_uri(&uri)
+    }
+
+    pub fn new_from_uri(uri: &URIReference<'_>) -> Result<Self, LocaterError> {
+        let scheme = uri.scheme().map(|s| s.as_str().to_ascii_lowercase());
+        let host = uri.host().map(|h| h.to_string());
+        match (scheme, host) {
+            (Some(scheme), Some(host)) if scheme == "usb" => {
+                let path = uri.path().segments().get(0).and_then(|s| {
+                    if !s.is_empty() {
+                        Some(s.as_str())
                     } else {
-                        return Err(DerivationPathError::InvalidDerivationPath(format!(
-                            "invalid query string `{}={}`, only `key` supported",
-                            pair.0, pair.1
-                        ))
+                        None
+                    }
+                });
+                let key = if let Some(query) = uri.query() {
+                    let parts = query
+                        .as_str()
+                        .split('?')
+                        .map(|c| {
+                            let mut it = c.split('=');
+                            let key = it.next().unwrap().to_string();
+                            let val = it.next().map(|v| v.to_string());
+                            (key, val)
```

Comment:
I guess it's not horribly long, but I don't love that this is manual now. Could we use `queryst`? Looks like it would simplify L404-L405 below

Reply:
I got the impression it has some odd quirks from somewhere.  I'll take a closer look though

Reply:
> I got the impression it has some odd quirks from somewhere. I'll take a closer look though

Oh right... `queryst` parses the string to JSON `Value` for some weird reason.

Reply:
I went with [`qstring`](https://crates.io/crates/qstring)

Reply:
Nice find! I was looking at `querystring`, but the helper methods on `qstring` are 🏅 .
This block turned out much more readable! ❤️